File: src/simulation/space/resource/dependency/AvoidDependency.java
package simulation.space.resource.dependency;

import simulation.space.resource.Resource;
import simulation.space.resource.tag.labeler.ResourceLabeler;
import simulation.space.tile.Tile;

import java.util.HashSet;
import java.util.Set;

public class AvoidDependency extends LabelerDependency {
    public Set<String> lastConsumed = new HashSet<>();

    public AvoidDependency(
            double amount,
            double deprivationCoefficient,
            boolean isNecessary,
            ResourceLabeler goodResources
    ) {
        super(deprivationCoefficient, isNecessary, amount, goodResources);
    }

    @Override
    double satisfaction(Tile tile, Resource resource) {
        double result;
        double _amount = amount * (resource == null ? 1 : resource.getAmount());
        int currentAmount = 0;
        for (Resource res : tile.getAccessibleResources()) {
            if (res.equals(resource)) {
                continue;
            }
            if (super.isResourceDependency(res)) {
                currentAmount += res.getAmount();
                if (currentAmount >= _amount) {
                    break;
                }
            }
        }
        result = Math.min(((double) currentAmount) / _amount, 1);
        return 1 - result;
    }

    @Override
    public boolean hasNeeded(Tile tile) {
        return tile.getAccessibleResources().stream().anyMatch(this::isResourceDependency);
    }

    @Override
    public boolean isResourceDependency(Resource resource) {
        return false;
    }

    @Override
    public boolean isPositive() {
        return false;
    }
}

File: src/simulation/space/resource/dependency/AvoidDependency.kt
package simulation.space.resource.dependency

import simulation.space.resource.Resource
import simulation.space.resource.tag.labeler.ResourceLabeler
import simulation.space.tile.Tile
import kotlin.math.min


class AvoidDependency(
        amount: Double,
        deprivationCoefficient: Double,
        isNecessary: Boolean,
        goodResources: ResourceLabeler
) : LabelerDependency(deprivationCoefficient, isNecessary, amount, goodResources) {
    var lastConsumed = mutableSetOf<String>()

    override fun satisfaction(tile: Tile, resource: Resource): Double {
        val result: Double
        val actualAmount = amount * resource.amount
        var currentAmount = 0

        for (list in tile.accessibleResources)
            for (res in list) {
                if (res == resource)
                    continue

                if (super.isResourceDependency(res)) {
                    currentAmount += res.amount * oneResourceWorth(res)
                    if (currentAmount >= actualAmount)
                        break
                }
            }
        result = min(currentAmount.toDouble() / actualAmount, 1.0)

        return 1 - result
    }

    override fun hasNeeded(tile: Tile) =
            tile.accessibleResources.any { it.asSequence().any { r -> isResourceDependency(r) } }

    override fun isResourceDependency(resource: Resource) = false

    override val isPositive = false
}
